src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -95,39 +96,106 @@ public TagHelperDescriptorProvider(IEnumerable<TagHelperDescriptor> descriptors)
 
         private IEnumerable<TagHelperDescriptor> ApplyRequiredAttributes(
             IEnumerable<TagHelperDescriptor> descriptors,
-            IEnumerable<string> attributeNames)
+            IEnumerable<KeyValuePair<string, string>> attributes)
         {
             return descriptors.Where(
                 descriptor =>
                 {
                     foreach (var requiredAttribute in descriptor.RequiredAttributes)
                     {
+                        var attributeName = GetRequiredAttributeName(requiredAttribute);
+                        IEnumerable<KeyValuePair<string, string>> matchingAttributes;
+
                         // '*' at the end of a required attribute indicates: apply to attributes prefixed with the
                         // required attribute value.
-                        if (requiredAttribute.EndsWith(
+                        if (attributeName.EndsWith(
                             RequiredAttributeWildcardSuffix,
                             StringComparison.OrdinalIgnoreCase))
                         {
-                            var prefix = requiredAttribute.Substring(0, requiredAttribute.Length - 1);
-
-                            if (!attributeNames.Any(
-                                attributeName =>
-                                    attributeName.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) &&
-                                    !string.Equals(attributeName, prefix, StringComparison.OrdinalIgnoreCase)))
-                            {
-                                return false;
-                            }
+                            var prefix = attributeName.Substring(0, attributeName.Length - RequiredAttributeWildcardSuffix.Length);
+                            matchingAttributes = attributes.Where(
+                                attribute => attribute.Key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) &&
+                                    !string.Equals(attribute.Key, prefix, StringComparison.OrdinalIgnoreCase));
                         }
-                        else if (!attributeNames.Contains(requiredAttribute, StringComparer.OrdinalIgnoreCase))
+                        else
+                        {
+                            matchingAttributes = attributes.Where(
+                                attribute => string.Equals(attribute.Key, attributeName, StringComparison.OrdinalIgnoreCase));
+                        }
+
+                        if (!matchingAttributes.Any())
                         {
                             return false;
                         }
+
+                        // At least one attribute where a name matches, now validate the value (if restricted).
+                        var equalsIndex = requiredAttribute.IndexOf('=');
```

Comment:
Ugly that the code to parse the attribute name is duplicated. Probably need a descriptor class to store the required attribute information when it's first discovered and validated.